django/thunderstore/repository/api/experimental/serializers.py

```diff
@@ -95,3 +96,132 @@ class Meta:
+        )
+
+    def to_representation(self, value):
+        return [c.pk for c in value]
+
+    def to_internal_value(self, data):
+        if not isinstance(data, list):
+            raise serializers.ValidationError("Not a list")
+
+        out = []
+        for category_pk in data:
+            category = PackageCategory.objects.filter(pk=category_pk).first()
+            if not category:
+                raise serializers.ValidationError(f"{category_pk} category not found")
+            out.append(category)
+        return out
+
+
+class PackageUploadSerializer(serializers.Serializer):
+    author_name = PackageUploadAuthorNameField()
+    categories = PackageUploadCategoriesField()
+    has_nsfw_content = serializers.BooleanField()
+    pk = serializers.IntegerField(read_only=True)
+    file = serializers.FileField(write_only=True)
+
+    def _create_form(self, data) -> PackageUploadForm:
+        request = self.context["request"]
+        return PackageUploadForm(
+            request.user,
+            data.get("author_name"),
+            request.community,
+            data={
+                "categories": data.get("categories"),
+                "has_nsfw_content": data.get("has_nsfw_content"),
+            },
+            files={"file": data.get("file")},
+        )
+
+    def validate(self, data):
+        form = self._create_form(data)
+        if not form.is_valid():
+            raise serializers.ValidationError("Invalid")
+        return data
+
+    def create(self, validated_data) -> PackageVersion:
+        form = self._create_form(validated_data)
+        form.is_valid()
+        return form.save()
+
+
+class UploaderIdentitySerializer(serializers.ModelSerializer):
+    class Meta:
+        model = UploaderIdentity
+        fields = (
+            "name",
+            "pk",
+        )
+
+    def create(self, *args, **kwargs):
+        raise Exception()
+
+    def update(self, *args, **kwargs):
+        raise Exception()
+
+
+class PackageSerializer(serializers.ModelSerializer):
+    owner = UploaderIdentitySerializer()
+
+    class Meta:
+        model = Package
+        fields = (
+            "owner",
+            "name",
+            "is_active",
+            "is_deprecated",
+            "date_created",
+            "date_updated",
+            "is_pinned",
+        )
+
+    def create(self, *args, **kwargs):
+        raise Exception()
+
+    def update(self, *args, **kwargs):
+        raise Exception()
+
+
+class PackageVersionSerializer(serializers.ModelSerializer):
+    package = PackageSerializer()
+
+    class Meta:
+        model = PackageVersion
+        fields = (
+            "package",
+            "is_active",
+            "date_created",
+            "downloads",
+            "name",
+            "version_number",
+            "website_url",
+            "description",
+            "dependencies",
+            "readme",
+            "file_size",
+            "pk",
+        )
```

Comment:
Could add a link to the icon's URL

Reply:
Good point, that exists in the v1 API

Reply:
Also exists in the `PackageVersionExperimentalSerializer`